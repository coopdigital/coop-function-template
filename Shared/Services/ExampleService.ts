import { Service } from 'typedi';
import ExampleRepository from '../Dal/ExampleRepository';
import { ExampleDto } from '../Dtos/ExampleDto';
import { ExampleModel } from '../Models/ExampleModel';

/** Example Service, abstract business logic from the Function */
@Service()
export default class ExampleService {
  /** Initialises the Service */
  constructor(private exampleRepository: ExampleRepository) {}

  /** Gets a example record by id */
  public async get(id: string): Promise<ExampleDto> {
    let response: ExampleDto = null;

    const exampleData = await this.exampleRepository.getExampleById(id);

    if (exampleData) {
      response = {
        Field1: exampleData.Field1,
        Field2: exampleData.Field2,
      };
    }

    return response;
  }

  /** Updates an example */
  public async update(id: string, example: ExampleDto): Promise<void> {
    const exampleModel: ExampleModel = { id, ...example };

    await this.exampleRepository.updateExampleModel(exampleModel);
  }
}
