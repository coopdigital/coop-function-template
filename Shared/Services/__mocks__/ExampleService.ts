export const mockGet = jest.fn((id) => {
  return {
    Field1: 'Test Data',
    Field2: 1234,
  };
});

export const mockUpdate = jest.fn();

const mock = jest.fn().mockImplementation(() => {
  return { get: mockGet, update: mockUpdate };
});
export default mock;
