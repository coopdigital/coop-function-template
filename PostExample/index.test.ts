import postExampleTrigger from './index';
import Context from '../testing/defaultContext';
// @ts-ignore
import { mockUpdate } from '../Shared/Services/ExampleService';
jest.mock('../Shared/Services/ExampleService.ts');
jest.mock('../Shared/Dal/ExampleRepository.ts');

describe('Post Example Tests', () => {
  beforeEach(() => {
    (mockUpdate as jest.Mock).mockClear();
  });

  it('Success', async () => {
    // Arrange
    const testContext = new Context();
    const testData = {
      params: {
        id: 'd8db7f54-7758-4a5a-bfd1-82e205c80a59',
      },
      body: {
        example: {
          Field1: 'Test Data',
          Field2: 1234,
        },
      },
    };

    // Act
    await postExampleTrigger(testContext, testData);

    // Assert
    expect((mockUpdate as jest.Mock).mock.calls.length).toBe(1);
    expect((mockUpdate as jest.Mock).mock.calls[0][0]).toBe(
      'd8db7f54-7758-4a5a-bfd1-82e205c80a59'
    );
    expect((mockUpdate as jest.Mock).mock.calls[0][1]).toMatchSnapshot();

    expect(testContext.res.status).toBe(200);
  });

  it('400 - Invalid ID', async () => {
    // Arrange
    const testContext = new Context();
    const testData = {
      params: {
        id: 'Test Id',
      },
      body: {
        example: {
          Field1: 'Test Data',
          Field2: 1234,
        },
      },
    };

    // Act
    await postExampleTrigger(testContext, testData);

    // Assert
    expect(testContext.res.status).toBe(400);
  });

  it('400 - Invalid Body', async () => {
    // Arrange
    const testContext = new Context();
    const testData = {
      params: {
        id: 'd8db7f54-7758-4a5a-bfd1-82e205c80a59',
      },
      body: {
        example: {
          Field3: 'Test Data',
          Field4: 1234,
        },
      },
    };

    // Act
    await postExampleTrigger(testContext, testData);

    // Assert
    expect(testContext.res.status).toBe(400);
  });

  it('500 - Service Exception', async () => {
    // Arrange
    (mockUpdate as jest.Mock).mockImplementationOnce(() => {
      throw new Error('Test Error');
    });

    const testContext = new Context();
    const testData = {
      params: {
        id: 'd8db7f54-7758-4a5a-bfd1-82e205c80a59',
      },
      body: {
        example: {
          Field1: 'Test Data',
          Field2: 1234,
        },
      },
    };

    // Act
    await postExampleTrigger(testContext, testData);

    // Assert
    expect((mockUpdate as jest.Mock).mock.calls.length).toBe(1);
    expect((mockUpdate as jest.Mock).mock.calls[0][0]).toBe(
      'd8db7f54-7758-4a5a-bfd1-82e205c80a59'
    );
    expect((mockUpdate as jest.Mock).mock.calls[0][1]).toMatchSnapshot();

    expect(testContext.res.status).toBe(500);
  });
});
