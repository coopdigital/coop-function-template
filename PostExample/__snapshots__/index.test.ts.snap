// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Example Tests 500 - Service Exception 1`] = `
Object {
  "Field1": "Test Data",
  "Field2": 1234,
}
`;

exports[`Post Example Tests Success 1`] = `
Object {
  "Field1": "Test Data",
  "Field2": 1234,
}
`;
